docs/HowSwiftImportsCModules.md

```diff
@@ -0,0 +1,922 @@
+Swift to import the C API, it is recommended to change C headers and to replace
+forward declarations with `#include`s of the header that defines the type.
+
+Incomplete types are often used intentionally to define opaque types. This is
+done too often, and Swift could not ignore this use case.  Swift imports
+pointers to incomplete types as `OpaquePointer`. For example:
+
+```c
+// C header.
+
+struct Foo;
+void Print(const Foo* foo);
+```
+
+```swift
+// C header imported in Swift.
+
+// Swift can't import the incomplete type `Foo` and has to drop some type
+// information when importing a pointer to `Foo`.
+func Print(_ foo: OpaquePointer)
+```
+
+# Function pointers
+
+C supports only one form of function pointers: `Result (*)(Arg1, Arg2, Arg3)`.
+
+Swift's closest native equivalent to a function pointer is a closure: `(Arg1,
+Arg2, Arg3) -> Result`. Swift closures don't have the same memory layout as C
+function pointers: closures in Swift consist of two pointers, a pointer to the
+code and a pointer to the captured data (the context). A C function pointer can
+be converted to a Swift closure; however, bridging is required to adjust the
+memory layout.
+
+As discussed above, there are cases where bridging that adjusts memory layout is
+not possible, for example, when importing pointers to function pointers. For
+example, while C's int `(*)(char)` can be imported as `(Int8) -> Int` (requires
+an adjustment of memory layout), C's `int (**)(char)` can't be imported as
+`UnsafePointer<(Int8) -> Int>`, because the pointee must have identical memory
+layout in C and in Swift.
+
+Therefore, we need a Swift type that has a memory layout identical to C function
+pointers, at least for such fallback cases. This type is spelled `@convention(c)
+(Arg1, Arg2, Arg3) -> Result`.
+
+Even though it is possible to import C function pointers as Swift closure with a
+context pointer in some cases, C function pointers are always imported as
+`@convention(c)` "closures" (in quotes because they don't have a context
+pointer, so they are not real closures). Swift provides an implicit conversion
+from `@convention(c)` closures to Swift closures with a context.
+
+Importing C function pointers also takes take pointer nullability into account:
+a nullable C function pointer is imported as optional.
+
+```c
+// C header.
+
+void qsort(
+  void *base,
+  size_t nmemb,
+  size_t size,
+  int (*compar)(const void *, const void *));
+
+void qsort_annotated(
+  void * _Nonnull base,
+  size_t nmemb,
+  size_t size,
+  int (* _Nonnull compar)(const void * _Nonnull, const void * _Nonnull));
+```
+
+```swift
+// C header imported in Swift.
+
+func qsort(
+  _ base: UnsafeMutableRawPointer!,
+  _ nmemb: Int32,
+  _ size: Int32,
+  _ compar: (@convention(c) (UnsafeRawPointer?, UnsafeRawPointer?) -> Int32)!
+)
+
+func qsort_annotated(
+  _ base: UnsafeMutableRawPointer,
+  _ nmemb: Int32,
+  _ size: Int32,
```

Comment:
```suggestion
  _ size: Int,
```

Reply:
Fixed.

Reply:
Not fixed. The four `size_t` parameters should be imported as `Int` rather than `CInt`.